### DISCUSSION TITLE
Overlay in a non-Debian distro

### DISCUSSION THREAD

**Discussion by Mogaba** [ltsp/ltsp#807] [Q&A] [answered]
I'm trying to use LTSP in a non-Debian distro (Alt Linux). This distro has it's own initramfs system with NFS root capabilitiy. I enabled it and used this cmdline to run LTSP (`/srv/ltsp/alt` is where I put the chroot):

`set cmdline_method root=/dev/nfs nfsroot=192.168.1.1:/srv/ltsp/alt,vers=3 init=/usr/share/ltsp/client/init/init`

It worked and the system booted but I think something went wrong during overlayfs mount:
```

$ findmnt
TARGET                       SOURCE              FSTYPE     OPTIONS
/                            /run/initramfs/ltsp overlay    rw,relatime,lowerdir=/,upperdir=/run/initramfs/ltsp/0/up,workdir=/run/initramfs/ltsp/0/work
├─/proc                      proc                proc       rw,nosuid,nodev,noexec,relatime
│ └─/proc/sys/fs/binfmt_misc systemd-1           autofs     rw,relatime,fd=29,pgrp=1,timeout=0,minproto=5,maxproto=5,direct,pipe_ino=15162
├─/dev                       udevfs              devtmpfs   rw,relatime,size=5120k,nr_inodes=500737,mode=755
│ ├─/dev/pts                 devpts              devpts     rw,nosuid,noexec,relatime,gid=5,mode=620,ptmxmode=000
│ ├─/dev/shm                 tmpfs               tmpfs      rw,nosuid,nodev
│ ├─/dev/hugepages           hugetlbfs           hugetlbfs  rw,relatime,pagesize=2M
│ └─/dev/mqueue              mqueue              mqueue     rw,nosuid,nodev,noexec,relatime
├─/sys                       sysfs               sysfs      rw,nosuid,nodev,noexec,relatime
│ ├─/sys/kernel/security     securityfs          securityfs rw,nosuid,nodev,noexec,relatime
│ ├─/sys/fs/cgroup           cgroup2             cgroup2    rw,nosuid,nodev,noexec,relatime,nsdelegate,memory_recursiveprot
│ ├─/sys/fs/pstore           pstore              pstore     rw,nosuid,nodev,noexec,relatime
│ ├─/sys/fs/bpf              none                bpf        rw,nosuid,nodev,noexec,relatime,mode=700
│ ├─/sys/kernel/debug        debugfs             debugfs    rw,nosuid,nodev,noexec,relatime
│ ├─/sys/kernel/tracing      tracefs             tracefs    rw,nosuid,nodev,noexec,relatime
│ ├─/sys/fs/fuse/connections fusectl             fusectl    rw,nosuid,nodev,noexec,relatime
│ └─/sys/kernel/config       configfs            configfs   rw,nosuid,nodev,noexec,relatime
└─/run                       runfs               tmpfs      rw,relatime,mode=755
  ├─/run/initramfs/ltsp      tmpfs               tmpfs      rw,relatime,mode=755
  ├─/run/user/0              tmpfs               tmpfs      rw,nosuid,nodev,noexec,relatime,size=402328k,nr_inodes=100582,mode=700
  └─/run/user/500            tmpfs               tmpfs      rw,nosuid,nodev,noexec,relatime,size=402328k,nr_inodes=100582,mode=700,uid=500,gid=500
```
As you can see the lower dir is /, the same as mountpoint, is it broken? I checked Ubuntu's overlay and the lower dir is /root there.

**Comment by alkisg** [comment, marked as answer]
Ubuntu's overlay happens in the initramfs stage, where the rootfs is mounted on a folder called /root. So there, lowerdir is /root.

Rasbperry Pi OS doesn't use an initramfs but supports nfs boot directly via its kernel. So there, /usr/share/ltsp/client/init/init detects that it's running in read-only media, and it adds the overlay. At that point, the rootfs is mounted on /. So lowerdir is /.

I think Alt Linux is the same as Rasbperry Pi OS. I even booted Fedora like that at some point, see previous discussions. So I believe that part went well.

The next issue you're likely to face is /etc/pam.d. Debian and Ubuntu use `pam-auth-update` to handle /etc/pam.d, and that's what LTSP uses as well. I don't know how Alt Linux manages its /etc/pam.d entries, but you need to update the LTSP PAM handling to cope with it. Otherwise SSH authentication and SSHFS home mounts won't work.

**Reply by Mogaba** [reply]
Thanks for the answer.

I found another issue, not sure if it's LTSP related. If I use persistent naming for network devices (ens18 in my case) system shutdown hangs with "A stop job is running for Network Connectivity". But if it's name is eth0, system shutdowns properly. I guess it has something to do with the fact that when the network goes down during the shutdown process, nfs root becomes unavailable?

**Comment by alkisg** [comment]
If you added persistent network names in /etc/udev or elsewhere, by hardcoding mac addresses, make sure these settings are excluded from the client image.

Other than that hint, I've no idea, I'd have to look into the Alt Linux initialization and shutdown scripts...

**Reply by Mogaba** [reply]
Could you please explain how nfs root is unmounted on shutdown?
I read that dracut does pivot_root back to the initramfs before the network goes down. I guess something similar happens in Ubuntu? Does LTSP take part in this process?

**Comment by alkisg** [comment]
No, LTSP doesn't take part in the shutdown process at all. It's all managed by distributions, systemd etc.
